x-pack/plugins/apm/server/lib/metrics/by_agent/shared/memory/index.ts

```diff
@@ -43,36 +45,97 @@ const chartBase: ChartBase = {
   series,
 };
 
-export const percentMemoryUsedScript = {
+/*
+  When no limit is specified in the container, docker allows the app as much memory / swap memory as it wants. 
+  This number represents the max possible value for the limit field.
+*/
+const CGROUP_LIMIT_MAX_VALUE = '9223372036854771712';
+
+export const percentSystemMemoryUsedScript = {
   lang: 'expression',
   source: `1 - doc['${METRIC_SYSTEM_FREE_MEMORY}'] / doc['${METRIC_SYSTEM_TOTAL_MEMORY}']`,
 };
 
+export const percentCgroupMemoryUsedScript = {
+  lang: 'painless',
+  source: `
+  boolean isFieldAvailable(def doc, def x) {
+    return doc.containsKey(x) 
+      && !doc[x].empty
+  }
+
+  if(!isFieldAvailable(doc, '${METRIC_CGROUP_MEMORY_USAGE_BYTES}')) {
+    return null;
+  }
+
+  double total = -1;
+  // uses cgroup.memory.mem.limit as total when it is available and not empty
+  if(isFieldAvailable(doc, '${METRIC_CGROUP_MEMORY_LIMIT_BYTES}')){
+    total = doc['${METRIC_CGROUP_MEMORY_LIMIT_BYTES}'].value;
+
+  //Otherwise uses system.memory.total as total
+  }else if (isFieldAvailable(doc, '${METRIC_SYSTEM_TOTAL_MEMORY}')){
+    total = doc['${METRIC_SYSTEM_TOTAL_MEMORY}'].value;
+  }
+  
+  // If both cgroup.memory.mem.limit and system.memory.total are not defined, does not calculate the percent and return null
+  if(total == -1) {
+    return null;
+
+  //When the cgroup limit is equal to CGROUP_LIMIT_MAX_VALUE, uses the system total to calculate
+  } else if(total == ${CGROUP_LIMIT_MAX_VALUE}L){
+    total = doc['${METRIC_SYSTEM_TOTAL_MEMORY}'].value;
+  }
+
+  double used = doc['${METRIC_CGROUP_MEMORY_USAGE_BYTES}'].value;
+  
+  return used / total;
+    `,
+};
```

Comment:
I think what we have is fine but I wanted to understand this better so tried to repro the issue.

I started by ingesting data with different mappings to different indices (indicies are created on write)
```jsonc
POST /customer/_doc
{
  "name": "Peter",
  "age": 30
}

POST /customer-2/_doc
{
  "name": "Susan",
  "hobby": "football"
}

POST /customer-3/_doc
{
  "name": "Anders",
  "from_age": 50,
  "to_age": 30
}
```

The following query fails:

```jsonc
GET customer*/_search
{
  "size": 0,
  "aggs": {
    "avg": {
      "avg": {
        "script": "1 - doc['from_age'].value - doc['to_age'].value"
      }
    }
  }
}
```

With this error:
```
          "script" : "1 - doc['from_age'].value - doc['to_age'].value",
          "lang" : "painless",
          "position" : {
            "offset" : 8,
            "start" : 0,
            "end" : 47
          },
          "caused_by" : {
            "type" : "illegal_argument_exception",
            "reason" : "No field found for [from_age] in mapping"
          }
```

But this query succeeds:

```jsonc
GET customer*/_search
{
  "size": 0,
  "query": {
    "bool": { "filter": [ { "exists": { "field": "from_age" } } ] }
  },
  "aggs": {
    "avg": {
      "avg": {
        "script": "1 - doc['from_age'].value - doc['to_age'].value"
      }
    }
  }
}
```

Reply:
The above examples are working like I would expect. So I'm not sure why the scripted fields we were using didn't also work (I assume that was the reason to move everything to painless?)